base/channels.jl

```diff
@@ -454,6 +470,155 @@ function put_unbuffered(c::Channel, v)
+"""
+function append!(c::Channel, iter::T) where {T}
+    if IteratorSize(iter) isa Union{HasLength, HasShape}
+        len = length(iter)
+        # shortcircuit for small short objects
+        if len == 0
+            return c
+        elseif len == 1
+            put!(c, @inbounds vec[begin])
+            return c
+        end
+    end
+
+    return isbuffered(c) ? append_buffered(c, iter) : append_unbuffered(c, iter)
+end
+function append!(c1::Channel, c2::Channel{T}) where {T}
+    # if either channel has a buffer size of 0 or 1, fallback to a naive `put!` loop
+    if min(c1.sz_max, c2.sz_max) < 2
+        return isbuffered(c1) ? append_buffered(c1, c2) : append_unbuffered(c1, c2)
+    end
+    i = 0
+    lock(c2)
+    try
+        while isopen(c2) || isready(c2)
+            # wait for data to become available
+            while isempty(c2.data)
+                do_is_closed() do
+                    check_channel_state(c2)
+                    wait(c2.cond_take)
+                end
+            end
+
+            lock(c1)
+            try
+                while length(c1.data) == c1.sz_max
+                    check_channel_state(c1)
+                    wait(c1.cond_put)
+                end
+
+                # append as many as possible to c1
+                n = min(c1.sz_max - length(c1.data), length(c2.data))
+                _increment_n_avail(c1, n)
+                c1_len_og = length(c1.data)
+
+                resize!(c1.data, c1_len_og + n)
+                copyto!(c1.data, c1_len_og + 1, c2.data, 1, n)
+                deleteat!(c2.data, 1:n)
+
+                _increment_n_avail(c2, -n)
+                notify(c1.cond_take, nothing, true, false)
+                foreach(_ -> notify(c2.cond_put, nothing, true, false), 1:n)
+            finally
+                unlock(c1)
+            end
+        end
+    finally
+        unlock(c2)
+    end
+    return c1
+end
+
+function append_unbuffered(c::Channel, iter)
+    for v in iter
+        put!(c, v)
+    end
+    return c
+end
+
+function append_buffered(c::Channel{T}, iter::I) where {T,I}
+    converted_items = Iterators.Stateful(Iterators.map(x -> convert(T, x), iter))
+
+    has_length = IteratorSize(iter) isa HasLength
+    elements_to_add = if has_length
+        # when we know the length, we can eagerly increment for all items
+        l = length(iter)
+        _increment_n_avail(c, l)
+        l
+    else
+        0
+    end
+
+    while !isempty(converted_items)
+        lock(c)
+        try
+            while length(c.data) == c.sz_max
+                check_channel_state(c)
+                wait(c.cond_put)
+            end
+
+            # Grab a chunk of items that will fit in the channel's buffer
+            available_space = c.sz_max - length(c.data)
+            chunk = Iterators.take(converted_items, available_space)
+
+            # for iterators without length, we increment the available items one item at a time
+            if !has_length
```

Comment:
I use this function as a fallbeck for appending an unbuffered channel to a buffered channel, but technically, we could use `n_avail` to figure out how many elements are ready to `take!` which would let us eagerly update n_avail for the `c_dst`. However, getting the channel-to-channel append right has been tricky, so I think that will need to be a second PR

Reply:
Actually, scratch that. Can't use `n_avail` because we might be competing with other consumers who could take some of the `n_avail`.